hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java

```diff
@@ -1404,15 +1404,26 @@ private JobTaskItemByHostMap generateTaskCountByHost(List<HostState> hosts, Iter
     }
 
     private List<JobTaskMoveAssignment> pushTasksOntoHost(HostState host, Iterable<HostState> heavyHosts) {
+        int moveLimit = config.getTasksMovedFullRebalance();
         MoveAssignmentList moveAssignments = new MoveAssignmentList(spawn, taskSizer);
+        List<JobTaskMoveAssignment> rv = purgeMisplacedTasks(host, moveLimit);
+
+        List<HostState> lightHostList = new ArrayList<>();
+        lightHostList.add(host);
+
+        double byteLimitFactor = 1;
+
         for (HostState heavyHost : heavyHosts) {
-            double byteLimitFactor =
-                    1 - ((double) moveAssignments.getBytesUsed() / config.getBytesMovedFullRebalance());
-            moveAssignments.addAll(pushTasksOffHost(heavyHost, Collections.singletonList(host), true, byteLimitFactor,
-                                                    config.getTasksMovedFullRebalance(), true));
+            if (moveAssignments.size() >= moveLimit || !canReceiveNewTasks(host) || moveAssignments.getBytesUsed() >= config.getBytesMovedFullRebalance()) {
+                break;
+            }
+            moveAssignments.addAll(pushTasksOffHost(heavyHost, lightHostList, true, byteLimitFactor, 1, true));
```

Comment:
Did you change the code back to moving multiple tasks from the same heavy host because of my comment on reusing code? I didn't mean reusing `pushTasksOffHost`, just the block of code that are duplicated between `pushTasksOffHost` and `pustTasksOntoHost` in your previous commit. I don't have a strong preference really. I think the idea of pulling tasks from different heave hosts has merit, but it's slightly more complex. I'm not convinced the benefit, if any, is worth that complexity.